### ISSUE TITLE
[BUG] [ws-sbom-generator] Missing "LicenseRef-" From Unknown License Names

### ISSUE THREAD

**Issue by benavidezb** [whitesource-ps/ws-sbom-generator#52]
According to SPDX Specification v2.2 Section 3.14.1, licenses not included in the SPDX License List (https://spdx.org/licenses/) should be preceded by "LicenseRef-". I have a project whose generated SBOM fails the Verify command from https://github.com/spdx/tools (can also use https://tools.spdx.org/app/validate/) with the following error:
_The following warning(s) were raised: [Package at line <###> invalid: Invalid license id 'Microsoft_.NET_Library'. Must start with 'LicenseRef-' and made up of the characters from the set 'a'-'z', 'A'-'Z', '0'-'9', '+', '_', '.', and '-'. in , < Package Name >]_

For some reason the Microsoft .NET Library license is not included in the SPDX license list and so it should be listed as "LicenseRef-Microsoft_.NET_Library" in SBOM reports.

**Comment by tamari-oz** [comment]
Thanks for the feedback!

I seperated SPDX licenses from extracted licenses and prefixed the latter with _LicenseRef-_.
It still gives an obscure error:
>  Invalid license id '__anon__d20c6021-4046-4efa-bb0d-1582bb03e379'.  Must start with 'LicenseRef-' and made up of the characters from the set 'a'-'z', 'A'-'Z', '0'-'9', '+', '_', '.', and '-'.

It might be a bug with the validation tool (both using the website and the jar verifier [here](https://github.com/spdx/tools-java)).
The tool reports other errors on missing fields which exist.

I generated pre-release (0.3.7a1) and if all's well I'll release GA soon.
`pip install ws-sbom-generator==0.3.7a1`

**Comment by benavidezb** [comment]
Thank you for the fast response! I installed the prerelease and I see that the hash algorithm string is updated but the Microsoft .NET Library license is still listed as "Microsoft_.NET_Library".

**Comment by tamari-oz** [comment]
Thanks for your feedback.

There is discrepancy whether '_' is allowed. 
According to the [spec](https://spdx.github.io/spdx-spec/other-licensing-information-detected/#1011-description) it's not:\
> LicenseRef-"[idstring]
where
[idstring] is a unique string containing letters, numbers, . and/or -.

So I added  regex that will replace anything that isn't A-Z, a-z, 0-9 '.' and '-'.
It will now replace the underscores in "Microsoft_.NET_Library" to "LicenseRef-Microsoft-.NET-Library".

Fix is in ws-sbom-generator 0.3.7a2.

**Comment by benavidezb** [comment]
Thanks for the update. Good find on underscores not being allowed by the spec. The validate test didn't even catch that. However, I tested 0.3.7a2 and I still see the license listed as "Microsoft_.NET_Library" with no "LicenseRef-". I see it listed next to PackageLicenseDeclared, PackageLicenseConcluded, and PackageLicenseInfoFromFiles for each package.

**Comment by benavidezb** [comment]
I tested 0.3.7 and the PackageLicenseDeclared/PackageLicenseConcluded for each .NET package is now listed as "NONE" instead of "Microsoft_.NET_Library". Example:

```
# Package

PackageName: runtime.aot.System.Globalization
SPDXID: SPDXRef-PACKAGE-runtime.aot.system.globalization.4.3.0.nupkg
PackageVersion: 4.3.0
PackageDownloadLocation: https://api.nuget.org/packages/runtime.aot.system.globalization.4.3.0.nupkg
FilesAnalyzed: False
PackageFileName: runtime.aot.system.globalization.4.3.0.nupkg
PackageSupplier: Organization: Microsoft (NOASSERTION)
PackageOriginator: Organization: Microsoft (NOASSERTION)
PackageChecksum: SHA1: 25d7fe78029f32a514106286782a39c8556c3708
PackageLicenseDeclared: NONE
PackageLicenseConcluded: NONE
```

Note also that PackageLicenseInfoFromFiles is removed.

I did notice that the Extracted Licenses section at the end of the SPDX document now includes a bunch of these (looks like one for each .NET package listed as "NONE"; however, there's no way to know which package they apply to):

```
LicenseID: LicenseRef-Microsoft-.NET-Library
LicenseName: LicenseRef-Microsoft-.NET-Library
ExtractedText: <text>Microsoft .NET Library</text>
```

**Comment by tamari-oz** [comment]
Thanks for the good feedback!

1. Regarding concluded licenses, this is a good point, according to SPDX 2.1 - Appendix IV, any SPDX expression should be fine including non-SPDX licenses - it will now do so (no support yet in composite licenses though).
2. I can see it is now populated (following fixing the concluded licenses).

I generated version 0.3.8a1 with a fix:  `pip install ws-sbom-generator==0.3.8a1`

**Comment by benavidezb** [comment]
That works, thanks!

```
# Package

PackageName: runtime.aot.System.Threading.Tasks
SPDXID: SPDXRef-PACKAGE-runtime.aot.system.threading.tasks.4.3.0.nupkg
PackageVersion: 4.3.0
PackageDownloadLocation: https://api.nuget.org/packages/runtime.aot.system.threading.tasks.4.3.0.nupkg
FilesAnalyzed: False
PackageFileName: runtime.aot.system.threading.tasks.4.3.0.nupkg
PackageSupplier: Organization: Microsoft (NOASSERTION)
PackageOriginator: Organization: Microsoft (NOASSERTION)
PackageChecksum: SHA1: 21ba6b918cef1e1b3b361795459b171fb46faefb
PackageLicenseDeclared: LicenseRef-Microsoft-.NET-Library
PackageLicenseConcluded: LicenseRef-Microsoft-.NET-Library
PackageLicenseInfoFromFiles: LicenseRef-Microsoft-.NET-Library
PackageCopyrightText: ['Copyright Microsoft Corporation']
PackageHomePage: https://api.nuget.org/packages/runtime.aot.system.threading.tasks.4.3.0.nupkg
```
